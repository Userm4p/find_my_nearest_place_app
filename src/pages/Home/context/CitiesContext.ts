import { createContext } from "react";
import { City } from "../types/cities";

export interface ICitiesContext {
  cities: City[];
  values: {
    city: string;
  };
  handleChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  reset: () => void;
  citiesToTable: City[];
  handleNextPage: () => void;
  handlePrevPage: () => void;
  pagesNumber: number;
  page: number;
}

export const CitiesContext = createContext<ICitiesContext>({
  cities: [],
  values: {
    city: "",
  },
  handleChange: () => {},
  reset: () => {},
  citiesToTable: [],
  handleNextPage: () => {},
  handlePrevPage: () => {},
  pagesNumber: 0,
  page: 0,
});
