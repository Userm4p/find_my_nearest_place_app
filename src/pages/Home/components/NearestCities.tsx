import { useContext, useMemo } from "react";
import { CitiesContext } from "../context/CitiesContext";
import { H2 } from "../../../components/H2/styled";
import { NearestCitiesEmpty, NearestCityItem, PageButton } from "./styled";
import { Colors } from "../../../utils/enums/Colors";

export const NearestCities = () => {
  const { nearestCities, selectedCity, clearSelectedCity } =
    useContext(CitiesContext);

  const isArrayEmpty = useMemo(
    () => nearestCities.length === 0,
    [nearestCities],
  );

  return (
    <div className="Home__NearestCities_container">
      <div className="Home__NearestCities_header">
        <H2 style={{
          color: Colors.Black
        }}>
          Selected city:{" "}
          <span style={{ fontWeight: "lighter" }}>
            {selectedCity?.name || "None"}
          </span>
        </H2>
        {selectedCity && (
          <PageButton data-testid="clear_button" onClick={clearSelectedCity}>
            Clear
          </PageButton>
        )}
      </div>
      {!isArrayEmpty ? (
        <div
          className="Home__NearestCities_cities_container"
          data-testid="nearest_citites"
        >
          {nearestCities.map((city, i) => (
            <NearestCityItem key={i}>
              <span
                style={{
                  fontWeight: "bold",
                  marginRight: "8px",
                }}
              >
                {i + 1}.
              </span>
              {city.name}
            </NearestCityItem>
          ))}
        </div>
      ) : (
        <NearestCitiesEmpty data-testid="empty_message">
          <span>You must select a city to find the nearest cities. ðŸ˜‰</span>
        </NearestCitiesEmpty>
      )}
    </div>
  );
};
