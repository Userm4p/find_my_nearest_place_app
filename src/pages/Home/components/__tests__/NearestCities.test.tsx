import { render } from "@testing-library/react";
import { NearestCities } from "../NearestCities";
import { CitiesContext } from "../../context/CitiesContext";
import "@testing-library/jest-dom";

const mockContext = {
  citiesToTable: Array.from({ length: 20 }, (_, i) => ({
    name: `City ${i}`,
    country: `Country ${i}`,
    lat: `${i}`,
    lng: `${i}`,
  })),
  pagesNumber: 20,
  page: 1,
  handleNextPage: () => {},
  handlePrevPage: () => {},
  values: { city: "" },
  handleChange: () => {},
  reset: () => {},
  cities: [],
  handleSelectCity: () => {},
  nearestCities: [],
  selectedCity: null,
  clearSelectedCity: () => {},
};

const mockNearestCities = Array.from({ length: 5 }, (_, i) => ({
  name: `City ${i}`,
  country: `Country ${i}`,
  lat: "i",
  lng: "i",
  distance: "i",
}));

describe("Tests in NearestCities", () => {
  test("should match snapshot when nearestCities is empty", () => {
    const { container, getByTestId } = render(
      <CitiesContext.Provider value={mockContext}>
        <NearestCities />
      </CitiesContext.Provider>,
    );
    expect(getByTestId("empty_message")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("should match snapshot when nearestCities is not empty", () => {
    const { container, getByTestId } = render(
      <CitiesContext.Provider
        value={{
          ...mockContext,
          nearestCities: mockNearestCities,
          selectedCity: mockNearestCities[0],
        }}
      >
        <NearestCities />
      </CitiesContext.Provider>,
    );
    expect(getByTestId("nearest_citites")).toBeInTheDocument();
    expect(getByTestId("clear_button")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
