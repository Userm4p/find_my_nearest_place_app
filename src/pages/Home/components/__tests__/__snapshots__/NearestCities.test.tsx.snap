// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in NearestCities should match snapshot when nearestCities is empty 1`] = `
<div>
  <div
    class="Home__NearestCities_container"
  >
    <div
      class="Home__NearestCities_header"
    >
      <h2
        class="sc-beySPh jhkGMy"
        style="color: rgb(0, 0, 0);"
      >
        Selected city:
         
        <span
          style="font-weight: lighter;"
        >
          None
        </span>
      </h2>
    </div>
    <div
      class="sc-eDLKkx ejAaeS"
      data-testid="empty_message"
    >
      <span>
        You must select a city to find the nearest cities. ðŸ˜‰
      </span>
    </div>
  </div>
</div>
`;

exports[`Tests in NearestCities should match snapshot when nearestCities is not empty 1`] = `
<div>
  <div
    class="Home__NearestCities_container"
  >
    <div
      class="Home__NearestCities_header"
    >
      <h2
        class="sc-beySPh jhkGMy"
        style="color: rgb(0, 0, 0);"
      >
        Selected city:
         
        <span
          style="font-weight: lighter;"
        >
          City 0
        </span>
      </h2>
      <button
        class="sc-dmyCSP iVBKLK"
        data-testid="clear_button"
      >
        Clear
      </button>
    </div>
    <div
      class="Home__NearestCities_cities_container"
      data-testid="nearest_citites"
    >
      <div
        class="sc-hLQSwg iBHetu"
      >
        <span
          style="font-weight: bold; margin-right: 8px;"
        >
          1
          .
        </span>
        City 0
      </div>
      <div
        class="sc-hLQSwg iBHetu"
      >
        <span
          style="font-weight: bold; margin-right: 8px;"
        >
          2
          .
        </span>
        City 1
      </div>
      <div
        class="sc-hLQSwg iBHetu"
      >
        <span
          style="font-weight: bold; margin-right: 8px;"
        >
          3
          .
        </span>
        City 2
      </div>
      <div
        class="sc-hLQSwg iBHetu"
      >
        <span
          style="font-weight: bold; margin-right: 8px;"
        >
          4
          .
        </span>
        City 3
      </div>
      <div
        class="sc-hLQSwg iBHetu"
      >
        <span
          style="font-weight: bold; margin-right: 8px;"
        >
          5
          .
        </span>
        City 4
      </div>
    </div>
  </div>
</div>
`;
